src/numerics/include/metaphysicl/numberarray.h

```diff
@@ -74,7 +74,7 @@ class NumberArray
     typedef NumberArray<N, T2> other;
   };
 
-  NumberArray() {}
+  NumberArray() : _data() {}
```

Comment:
At the risk of revealing my own ignorance: why is this necessary?

Is the idea to force _data to be initialized rather than garbage?  At the risk of undermining my above rant about safety, I must admit that allowing uninitialized data where possible was a deliberate design decision.  I (albeit by necessity rather than choice) trust C/C++ users not to expect that "some_numeric_type n;" will be zero-initialized.

My decision is debatable, I admit.

Reply:
This was a change I implemented a long time ago (a month?) when I was adapting idaholab/moose#5661 to modern MOOSE. I was getting compiler warnings about possible use of uninitialized data, which was as far as I can tell, a totally false positive... But since we don't allow compiler warnings in MOOSE, I made this change.

Reply:
I could leave this out of this PR, and return to it later when we are ready to merge AD into MOOSE

Reply:
Oh, hell, there's no way to allow ```NumArr n;``` to remain undefined without also preventing ```NumArr n = NumArr();``` from returning a zeroed array, is there?  In that case returning initialized data in both cases would definitely be preferable to returning uninitialized data in both.  If I wasn't in the habit of zero-initializing via ```NumArr n = 0;``` this would have bitten me years ago, and it could bite any other user writing generic code with the other idiom.

Reply:
If I've understood what's going on here then let me know so I can open up an issue; NumberArray isn't the only class with this problem.

Reply:
Leave this in this PR, unless you see a way I've missed to mimic built-in behavior more properly.  I would still like it if users had the ability to leave data uninitialized via constructor idioms that are expected do so on purpose, but it was a grotesque mistake that I made it so easy for users to leave data uninitialized when using idioms that are expected to zero-initialize it.